modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EBinary.java

```diff
@@ -280,6 +285,20 @@ private void analyzeSub(Locals variables) {
         actual = promote;
     }
 
+    private void analyzeRegexOp(Locals variables) {
+        left.analyze(variables);
+        right.analyze(variables);
+
+        left.expected = Definition.STRING_TYPE;
+        right.expected = Definition.PATTERN_TYPE;
+
+        left = left.cast(variables);
+        right = right.cast(variables);
+
+        // It'd be nice to be able to do constant folding here but we can't because constants aren't flowing through EChain
```

Comment:
Definitely for a different PR, but I think we can fix this by making the String constant in ALink an Object constant instead.